### ISSUE TITLE
3.6.0 (darkroom) unstable on Gentoo Linux

### ISSUE THREAD

**Issue by frli8848** [darktable-org/darktable#9815]
I get random crashes when editing Nikon .NEF files in 3.6.0 I started clean by deleting:

<code>
.config/darktable/data.db

.config/darktable/library.db
</code>

I just get random segfaults when pressing "d" in lightable. Restarting darktable and opening another file first sometimes help but not always.

Also, applying a mask to, at least the exposure module, always results in:
<code>
darktable: /var/tmp/portage/media-gfx/darktable-3.6.0/work/darktable-3.6.0/src/develop/pixelpipe_hb.c:1470: dt_dev_pixelpipe_process_rec: Assertion `tiling.factor > 0.0f' failed.
Aborted
</code>
That is, doing:
1. mask manager -> create mask
2. exposure module  -> new instance
3. exposure module -> drawn mask
4. exposure module -> select the created mask
5. Abort

I had similar problems with masks in 3.4.1 when the mask extended outside the image but on 3.6.0 it always aborts darktable.

**Comment by frli8848** [comment]
I did some more testing and the mask manager sort-of-works if one adds another shape, a circle for example, and then select the drawn mask (path). That is, it crashes less often if the added shape is not last in the list of shapes it seems. And it also helps to go back to the lightable and then back to darkroom so that the changes are written to disk before it crashes...

**Comment by johnny-bit** [comment]
Hi,

Your report has not enough data and can't replicate your problem following outlined steps (and I do have gentoo). Can you attach crash logs, xmp files and your system description (ram, opencl etc)?

**Comment by frli8848** [comment]
Here is my system info:

```
# emerge --info

Portage 3.0.20 (python 3.9.6-final-0, default/linux/amd64/17.1/desktop/plasma, gcc-11.2.0, glibc-2.33-r1, 5.12.19-gentoo x86_64)
=================================================================
System uname: Linux-5.12.19-gentoo-x86_64-AMD_Ryzen_9_5950X_16-Core_Processor-with-glibc2.33
KiB Mem:    65832052 total,   3759812 free
KiB Swap:          0 total,         0 free
Timestamp of repository gentoo: Mon, 16 Aug 2021 06:00:01 +0000
Head commit of repository gentoo: e7c78f9f71f2cbc13def16ceac7758b2c2802e13
sh bash 5.1_p8
ld GNU ld (Gentoo 2.35.2 p1) 2.35.2
app-shells/bash:          5.1_p8::gentoo
dev-java/java-config:     2.3.1::gentoo
dev-lang/perl:            5.34.0::gentoo
dev-lang/python:          2.7.18_p11::gentoo, 3.6.14::gentoo, 3.7.11::gentoo, 3.8.11::gentoo, 3.9.6_p1::gentoo
dev-lang/rust:            1.54.0::gentoo
dev-util/cmake:           3.21.1::gentoo
sys-apps/baselayout:      2.7::gentoo
sys-apps/openrc:          0.42.1-r1::gentoo
sys-apps/sandbox:         2.24::gentoo
sys-devel/autoconf:       2.13-r1::gentoo, 2.69-r5::gentoo
sys-devel/automake:       1.13.4-r2::gentoo, 1.16.3-r1::gentoo
sys-devel/binutils:       2.35.2::gentoo
sys-devel/gcc:            9.3.0-r2::gentoo, 10.3.0-r2::gentoo, 11.2.0::gentoo
sys-devel/gcc-config:     2.4::gentoo
sys-devel/libtool:        2.4.6-r6::gentoo
sys-devel/make:           4.3::gentoo
sys-kernel/linux-headers: 5.12::gentoo (virtual/os-headers)
sys-libs/glibc:           2.33-r1::gentoo
Repositories:

gentoo
    location: /var/db/repos/gentoo
    sync-type: rsync
    sync-uri: rsync://rsync.gentoo.org/gentoo-portage
    priority: -1000
    sync-rsync-verify-jobs: 1
    sync-rsync-extra-opts: 
    sync-rsync-verify-max-age: 24
    sync-rsync-verify-metamanifest: yes


dotnet
    location: /var/lib/layman/dotnet
    masters: gentoo
    priority: 50


ACCEPT_KEYWORDS="amd64"
ACCEPT_LICENSE="@FREE"
CBUILD="x86_64-pc-linux-gnu"
CFLAGS="-O3 -pipe -march=native -mtune=native"
CHOST="x86_64-pc-linux-gnu"
CONFIG_PROTECT="/etc /usr/lib64/libreoffice/program/sofficerc /usr/share/config /usr/share/gnupg/qualified.txt"
CONFIG_PROTECT_MASK="/etc/ca-certificates.conf /etc/dconf /etc/env.d /etc/fonts/fonts.conf /etc/gconf /etc/gentoo-release /etc/php/apache2-php7.4/ext-active/ /etc/php/cgi-php7.4/ext-active/ /etc/php/cli-php7.4/ext-active/ /etc/revdep-rebuild /etc/sandbox.d /etc/terminfo /etc/texmf/language.dat.d /etc/texmf/language.def.d /etc/texmf/updmap.d /etc/texmf/web2c"
CXXFLAGS="-O3 -pipe -march=native -mtune=native"
DISTDIR="/var/cache/distfiles"
ENV_UNSET="CARGO_HOME DBUS_SESSION_BUS_ADDRESS DISPLAY GOBIN GOPATH PERL5LIB PERL5OPT PERLPREFIX PERL_CORE PERL_MB_OPT PERL_MM_OPT XAUTHORITY XDG_CACHE_HOME XDG_CONFIG_HOME XDG_DATA_HOME XDG_RUNTIME_DIR"
FCFLAGS="-O3 -pipe -march=native -mtune=native"
FEATURES="assume-digests binpkg-docompress binpkg-dostrip binpkg-logs config-protect-if-modified distlocks ebuild-locks fixlafiles ipc-sandbox merge-sync multilib-strict network-sandbox news parallel-fetch pid-sandbox preserve-libs protect-owned qa-unresolved-soname-deps sandbox sfperms strict unknown-features-warn unmerge-logs unmerge-orphans userfetch userpriv usersandbox usersync xattr"
FFLAGS="-O3 -pipe -march=native -mtune=native"
GENTOO_MIRRORS="http://distfiles.gentoo.org"
LDFLAGS="-Wl,-O1 -Wl,--as-needed"
MAKEOPTS="-j16"
PKGDIR="/var/cache/binpkgs"
PORTAGE_CONFIGROOT="/"
PORTAGE_RSYNC_OPTS="--recursive --links --safe-links --perms --times --omit-dir-times --compress --force --whole-file --delete --stats --human-readable --timeout=180 --exclude=/distfiles --exclude=/local --exclude=/packages --exclude=/.git"
PORTAGE_TMPDIR="/var/tmp"
USE="X a52 aac acl acpi activities alsa amd64 bluetooth branding bzip2 cairo cdda cddb cdparanoia cdr cli colord crypt cups dbus declarative dri dts dvd dvdr elogind emboss encode eselect-ldso exif flac fortran gdbm gif gphoto2 gpm gtk gui iconv icu ipv6 jack jpeg kde kipi kwallet ladspa lcms libglvnd libnotify libtirpc lm_sensors mad mng mp3 mp4 mpeg mtp multilib ncurses nls nptl ogg opengl openmp pam pango pcre pdf phonon plasma png policykit ppds qml qt5 readline sdl seccomp semantic-desktop spell split-usr ssl startup-notification svg tcpd tiff truetype twolame udev udisks unicode upower usb vdpau vorbis widgets wxwidgets x264 xattr xcb xml xv xvid zlib" ABI_X86="64" ADA_TARGET="gnat_2018" ALSA_CARDS="hdsp" APACHE2_MODULES="authn_core authz_core socache_shmcb unixd actions alias auth_basic authn_alias authn_anon authn_dbm authn_default authn_file authz_dbm authz_default authz_groupfile authz_host authz_owner authz_user autoindex cache cgi cgid dav dav_fs dav_lock deflate dir disk_cache env expires ext_filter file_cache filter headers include info log_config logio mem_cache mime mime_magic negotiation rewrite setenvif speling status unique_id userdir usertrack vhost_alias" CALLIGRA_FEATURES="karbon sheets words" COLLECTD_PLUGINS="df interface irq load memory rrdtool swap syslog" CPU_FLAGS_X86="aes avx avx2 f16c fma3 mmx mmxext pclmul popcnt sha sse sse2 sse3 sse4_1 sse4_2 sse4a ssse3" ELIBC="glibc" GPSD_PROTOCOLS="ashtech aivdm earthmate evermore fv18 garmin garmintxt gpsclock greis isync itrax mtk3301 nmea ntrip navcom oceanserver oldstyle oncore rtcm104v2 rtcm104v3 sirf skytraq superstar2 timing tsip tripmate tnt ublox ubx" INPUT_DEVICES="libinput" KERNEL="linux" L10N="eu no sv en" LCD_DEVICES="bayrad cfontz cfontz633 glk hd44780 lb216 lcdm001 mtxorb ncurses text" LIBREOFFICE_EXTENSIONS="presenter-console presenter-minimizer" LUA_SINGLE_TARGET="lua5-1" LUA_TARGETS="lua5-1" OFFICE_IMPLEMENTATION="libreoffice" PHP_TARGETS="php7-3 php7-4" POSTGRES_TARGETS="postgres11 postgres13" PYTHON_SINGLE_TARGET="python3_9" PYTHON_TARGETS="python3_7 python3_8 python3_9" RUBY_TARGETS="ruby26" USERLAND="GNU" VIDEO_CARDS="nvidia" XTABLES_ADDONS="quota2 psd pknock lscan length2 ipv4options ipset ipp2p iface geoip fuzzy condition tee tarpit sysrq proto steal rawnat logmark ipmark dhcpmac delude chaos account"
Unset:  CC, CPPFLAGS, CTARGET, CXX, EMERGE_DEFAULT_OPTS, INSTALL_MASK, LANG, LC_ALL, LINGUAS, PORTAGE_BINHOST, PORTAGE_BUNZIP2_COMMAND, PORTAGE_COMPRESS, PORTAGE_COMPRESS_FLAGS, PORTAGE_RSYNC_EXTRA_OPTS, RUSTFLAGS
```

Graphics; Nvidia 3090 Ti 
```
# eix -s nvidia-drivers
[I] x11-drivers/nvidia-drivers
     Available versions:  390.144(0/390) 460.91.03(0/460) 470.57.02(0/470) (~)470.63.01(0/470) {+X dist-kernel (+)driver static-libs +tools wayland ABI_MIPS="n32 n64 o32" ABI_S390="32 64" ABI_X86="32 64 x32" KERNEL="linux"}
     Installed versions:  470.63.01(0/470)(09:18:02 08/12/21)(X driver static-libs tools -dist-kernel -wayland ABI_MIPS="-n32 -n64 -o32" ABI_S390="-32 -64" ABI_X86="64 -32 -x32" KERNEL="linux")
     Homepage:            https://www.nvidia.com/download/index.aspx
     Description:         NVIDIA Accelerated Graphics Driver
```

```
$ nvidia-smi 
Sun Aug 29 22:07:29 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.63.01    Driver Version: 470.63.01    CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  Off  | 00000000:0A:00.0  On |                  N/A |
|  0%   52C    P8    48W / 350W |   2090MiB / 12050MiB |      6%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

```

This is the most unstable version I have tested so far and I get frequent crashes by just pressing the d key in light table. I will try to build 3.6.0 from sources with debug symbols to see if I can get better traces of crashes.

**Comment by johnny-bit** [comment]
And this is why we need backtraces:

> ```
> 0x00007fff8c701737 in pixel_correction (exposure=<optimized out>, factors=factors@entry=0x55555a759b10, sigma=sigma@entry=1.41421354)
>    at /home/fl/software/darktable-3.6.0/src/iop/toneequal.c:783
>783         result += gaussian_func(expo - centers_ops[i], gauss_denom) * factors[i];
>```

because now we can correlate it with #9340 and #9002 and https://github.com/darktable-org/darktable/issues/8695#issuecomment-860141838

Most important thing here: https://github.com/darktable-org/darktable/issues/9002#issuecomment-871090193

Current master does not exhibit crashes and there are no crashes when compilled with GCC10.

The problem is somewhere in GCC probably, I opened a bug report about it here: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=101262

If you could create reproducer it would be awesome.

**Comment by frli8848** [comment]
For reference, since I have gcc-10.3 and clang-12 on the machine too:

First try gcc-10.3.0
```
 # gcc-config -l
 [1] x86_64-pc-linux-gnu-9.3.0
 [2] x86_64-pc-linux-gnu-10.3.0
 [3] x86_64-pc-linux-gnu-11.2.0 *
 # gcc-config 2 
 * Switching native-compiler to x86_64-pc-linux-gnu-10.3.0 ...
>>> Regenerating /etc/ld.so.cache...                                                                                                                       [ ok ]

 * If you intend to use the gcc from the new profile in an already
 * running shell, please remember to do:

 *   . /etc/profile
```

```
~/software/darktable-3.6.0/build $ . /etc/profile
~/software/darktable-3.6.0/build $ cmake -DCMAKE_C_FLAGS="-O3 -march=native" -DCMAKE_CXX_FLAGS="-O3 -march=native" -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=/usr -DBUILD_CURVE_TOOLS=no -DBUILD_NOISE_TOOLS=no -DBUILD_PRINT=yes -DCUSTOM_CFLAGS=ON -DDONT_USE_INTERNAL_LUA=ON -DRAWSPEED_ENABLE_LTO=no -DTESTBUILD_OPENCL_PROGRAMS=OFF -DUSE_AVIF=no -DUSE_CAMERA_SUPPORT=yes -DUSE_COLORD=yes -DUSE_FLICKR=no -DUSE_GMIC=no -DUSE_GRAPHICSMAGICK=yes -DUSE_KWALLET=yes -DUSE_LIBSECRET=no -DUSE_LUA=no -DUSE_MAP=yes -DUSE_NLS=yes -DUSE_OPENCL=yes -DUSE_OPENEXR=yes -DUSE_OPENJPEG=yes -DUSE_OPENMP=yes -DUSE_WEBP=yes -DWANT_JSON_VALIDATION=no -DCMAKE_BUILD_TYPE=RELWITHDEBINFO ..
~/software/darktable-3.6.0/build $ make -j16

```
This builds and runs fine. Got no crashes as with gcc-11 and -march=native.

Next try clang-12

```
~/software/darktable-3.6.0/build $ clang --version
clang version 12.0.1
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/lib/llvm/12/bin

~/software/darktable-3.6.0/build $ cmake -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang -DCMAKE_LINKER=ld.ldd  -DCMAKE_C_FLAGS="-O3 -march=native" -DCMAKE_CXX_FLAGS="-O3 -march=native" -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=/usr -DBUILD_CURVE_TOOLS=no -DBUILD_NOISE_TOOLS=no -DBUILD_PRINT=yes -DCUSTOM_CFLAGS=ON -DDONT_USE_INTERNAL_LUA=ON -DRAWSPEED_ENABLE_LTO=no -DTESTBUILD_OPENCL_PROGRAMS=OFF -DUSE_AVIF=no -DUSE_CAMERA_SUPPORT=yes -DUSE_COLORD=yes -DUSE_FLICKR=no -DUSE_GMIC=no -DUSE_GRAPHICSMAGICK=yes -DUSE_KWALLET=yes -DUSE_LIBSECRET=no -DUSE_LUA=no -DUSE_MAP=yes -DUSE_NLS=yes -DUSE_OPENCL=yes -DUSE_OPENEXR=yes -DUSE_OPENJPEG=yes -DUSE_OPENMP=yes -DUSE_WEBP=yes -DWANT_JSON_VALIDATION=no -DCMAKE_BUILD_TYPE=RELWITHDEBINFO ..
~/software/darktable-3.6.0/build $ make -j16

-snip-

[ 31%] Building C object bin/CMakeFiles/lib_darktable.dir/common/imageio_rgbe.c.o
/usr/bin/x86_64-pc-linux-gnu-ld: ../../../librawspeed.a(RawImage.cpp.o): undefined reference to symbol '_ZTVN10__cxxabiv121__vmi_class_type_infoE@@CXXABI_1.3'
/usr/bin/x86_64-pc-linux-gnu-ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.2.0/libstdc++.so.6: error adding symbols: DSO missing from command line
clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [lib64/darktable/rawspeed/src/utilities/identify/CMakeFiles/darktable-rs-identify.dir/build.make:105: lib64/darktable/rawspeed/src/utilities/identify/darktable-rs-identify] Error 1
make[1]: *** [CMakeFiles/Makefile2:1935: lib64/darktable/rawspeed/src/utilities/identify/CMakeFiles/darktable-rs-identify.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

Some LDFLAGS fixes this

```
~/software/darktable-3.6.0/build $ rm -fR *
~/software/darktable-3.6.0/build $ export LDFLAGS="-Wl,--copy-dt-needed-entries" 
~/software/darktable-3.6.0/build $ cmake -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang -DCMAKE_LINKER=ld.ldd  -DCMAKE_C_FLAGS="-O3 -march=native" -DCMAKE_CXX_FLAGS="-O3 -march=native" -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=/usr -DBUILD_CURVE_TOOLS=no -DBUILD_NOISE_TOOLS=no -DBUILD_PRINT=yes -DCUSTOM_CFLAGS=ON -DDONT_USE_INTERNAL_LUA=ON -DRAWSPEED_ENABLE_LTO=no -DTESTBUILD_OPENCL_PROGRAMS=OFF -DUSE_AVIF=no -DUSE_CAMERA_SUPPORT=yes -DUSE_COLORD=yes -DUSE_FLICKR=no -DUSE_GMIC=no -DUSE_GRAPHICSMAGICK=yes -DUSE_KWALLET=yes -DUSE_LIBSECRET=no -DUSE_LUA=no -DUSE_MAP=yes -DUSE_NLS=yes -DUSE_OPENCL=yes -DUSE_OPENEXR=yes -DUSE_OPENJPEG=yes -DUSE_OPENMP=yes -DUSE_WEBP=yes -DWANT_JSON_VALIDATION=no -DCMAKE_BUILD_TYPE=RELWITHDEBINFO ..
~/software/darktable-3.6.0/build $ make -j16
```

This also seems to run fine.

**Comment by johnny-bit** [comment]
Sooo... It looks like your report is regarding issue that is fixed in newest 11 series GCC, not present before or after 11.x and not present in LLVM. I'd say this is fixed and/or workaround is easily "present"

**Comment by frli8848** [comment]
@johnny-bit Do you have a link to where this fix for gcc 11.2.0 is discussed? Just curious since I have issues with gcc-10.3.0 too (internal compiler error when building LibreOffice)

**Comment by johnny-bit** [comment]
I've opened 2 bugs against gcc.
Solved one: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=101261
Waiting for reproducers: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=101262

Reproducers for bugs are very important :)
